### PR TITLE
Domain Management: add expired domain notice

### PR DESCRIPTION
**Pull request by jeremeylduvall**
Potential fix for #2422:

- Adds a check to see if the domain is expiring or expired (returns null if type is `WPCOM`)
- If the domain is expired, displays notice in the form of "Expired 3 days ago"
- If the domain is expiring in the next 30 days, displays notice in the form of "Expires in 2 days"

It's not quite ready to go just yet. I'm running into an issue when the domain is expiring in the near term (within a few hours). On the Purchases page, the `purchase.expiryMoment` comes in the form `"2016-01-27T00:00:00+00:00"`. On Domain Management, `domain.expirationMoment` doesn't have a time component (comes like `"January 25, 2016"`). I believe this is why the time is rendering differently. It's always a 7 hour time difference (I'm GMT-7 timezone). 

Here's the Purchases view of a domain set to expire today (Reads "Expired 4 hours ago"):

screen shot 2016-01-24 at 9 09 33 pm

Here's that same domain from Domain Management. It recognizes as expired, but it reads "Expired in 3 hours":

screen shot 2016-01-24 at 9 09 15 pm

Is there a simple way around this?

Any and all feedback welcome. I borrowed the `domain.expirationMoment` logic from `purchases/list/item/index.jsx`.

### CODE REVIEW

File: client/my-sites/upgrades/domain-management/list/item.jsx [hunk 1]
```diff
@@ -32,6 +33,38 @@ const ListItem = React.createClass( {
...
+				</Notice>
+			);
+		}
+
+		if ( domain.expirationMoment < this.moment().add( 30, 'days' ) ) {
+			return (
+				<Notice isCompact status="is-error" icon="spam">
+					{ this.translate( 'Expires %(timeUntilExpiry)s', {
+						args: {
+							timeUntilExpiry: domain.expirationMoment.fromNow()
```
Comment (drewblaisdell) [inline client/my-sites/upgrades/domain-management/list/item.jsx]: `moment#fromNow()` returns a string representing a human-readable interval of time, ranging from "[years] years ago" to "a few seconds ago". The expiry for a given domain, however, is returned from the API without hours/minutes/seconds, so I think you want to avoid dealing with units of < 1 day here.

The way we handle this when displaying this information for free trials is to get the difference between the expiry date (from the start of the day, which we calculate with [moment#startOf( 'day' )](https://github.com/Automattic/wp-calypso/blob/master/client/state/sites/plans/assembler.js#L11)) and the current date (from the start of the day) in days. You can calculate the difference in days between two moments like this:

```js
domain.expirationMoment.diff( this.moment(), 'days' );
```

We can check from the start of each day like this:

```js
domain.expirationMoment.startOf( 'day' ).diff( this.moment().startOf( 'day' ), 'days' );
```

Then we just display whatever value that returns in a translate string, e.g. `'Expired %(daysAfterExpiry)s days ago'.

In the case where the API still returns a true `expired` value but there are zero days remaining until expiry, we handle this case with its own string in `PlanProgress`: 'Expires momentarily'

**Comment by jeremeylduvall** [comment]
Made some adjustments and removed the use of `fromNow()`. Instead, used `domain.expirationMoment.startOf( 'day' ).diff( this.moment().startOf( 'day' ), 'days' );` to return the difference between the expiration date and the current date in whole number form. 

First, the conditional `this.props.domain.type !== 'WPCOM' ? this.isDomainExpired( this.props.domain ) : null` only runs the `.isDomainExpired()` function if the domain type is not WPCOM.

Inside `.isDomainExpired()`, the function checks four specific cases with the variable `expirationTime`, which is the difference between the current date (start of the day) and the expiration date (start of the day):

1. Is `domain.expired` true...
A) and `expirationTime` between -1 and 1. If so, it returns 'Expired a day ago'
B) If not, it returns 'Expired %(timeSinceExpiry)s days ago' where `timeSinceExpiry` is the opposite of `expirationTime` (since it returns negative)

2. Is `expirationTime` < 30... 
A) and equal to 0 or 1. If so, it returns 'Expires today'
B) If not, it returns 'Expires in %(timeUntilExpiry)s days' where `timeUntilExpiry` is equal to `expirationTime` (a positive number always > 1)

Screenshot in action:

screen shot 2016-01-31 at 9 19 13 pm

Feels a bit wordy, but I was struggling to deal with all four cases (expiring or expired short term vs. long term) without the current setup.

**Comment by umurkontaci** [comment]
Can we get around this by using `utc` mode in moment and setting the utc offset to registrar's timezone? I'm investigating at what time exactly registrar's expire their domains.
http://momentjs.com/docs/#/parsing/utc/

**Comment by umurkontaci** [comment]
I've found out that the actual expiration time is "exactly" one year past the registration date for a one-year registration (excluding server/registry delays). We should have an accurate record of the registration timestamp down to the minute, so I think we should be able to make something out of it. 

We can return a proper ISO8601 date string as with domain for registration timestamp, but we somehow our expiration records have only down-to-day precision (no hour/minute), but we might be able to deduce it from registration date.

**Comment by jeremeylduvall** [comment]
@umurkontaci take a look at the most recent commit? Seems like the main issue is that `domain.expirationMoment` in `domain-management/list/item.jsx` is including a GMT offset for each user. 

For example, I set a domain mapping to expire today. When I log out `domain.expirationMoment.toISOString()`, I get `2016-02-22T07:00:00.000Z` (I'm GMT-7). Meanwhile, for the same domain upgrade `purchases/list/item/index.jsx` is returning `2016-02-22T00:00:00.000Z`.

Within the `isDomainExpired` function, I created a new variable `domainExpirationDate` and used `.utc().startOf('day')` like you mentioned to set `domain.expirationMoment`, which matches what is returned from `purchases/list/item/index.jsx`. Now, `fromNow()` is giving me the correct value.

This simplifies things quite a bit since now we only have to worry about two cases (expired and expiring within 30 days).

**Comment by jeremeylduvall** [comment]
Updated this one more time. The only change is in the comparison for domains expiring within 30 days. I changed the comparison moment to `this.moment().add( 30, 'days' ).startOf( 'day' )` using the `startOf()` function so that it matches the conditions we're comparing against (`domainExpirationDate`).

If we want to move forward with this fix, I think this should be good to go. To test:

1. Add some expiring and expired domains to your site. I setup 5 domains: 1/2 days past expiration, Expiring today, and 1/2 days prior to expiration 
2. Visit wordpress.com/purchases. We already render notices there so now, we have something to compare against.
3. Visit the domains page (https://wordpress.com/domains/manage/sitename.wordpress.com) for the specific site. The notices should match step #2.

Here's the purchases page with the domains I tested:

screen shot 2016-02-24 at 7 27 06 am

Here's the domains page (notices should match):

screen shot 2016-02-24 at 7 26 28 am

**Comment by umurkontaci** [comment]
@schwuk I think it's a good idea to fix this properly by returning a correct ISO8601 string with an updated expiration date string. (Currently it's set to start of day so hour/minute detail is removed – which should be the same as registration timestamp)

Do you think you Lava can do this?

**Comment by klimeryk** [comment]
It seems @aidvu did this change (`expiry` in ISO8601) in `D1382-code` - can you give this PR another try, @jeremeylduvall? :) `domain.expirationMoment` should have [all the info you need](https://github.com/Automattic/wp-calypso/blob/master/client/lib/domains/assembler.js#L25) (introduced in #4194).
Along with #4256 it would make it so much harder to accidentally loose a domain.

**Comment by jeremeylduvall** [comment]
@klimeryk thanks! This is much simpler with `domain.expirationMoment`. Still using the same logic mentioned above checking for `domain.expired` and `domain.expirationMoment < this.moment().add( 30, 'days' )`. Still works as mentioned previously. Here's a screenshot of `/domains/manage` on the branch compared with `/me/purchases`. Timeframes are matching up. If this is on the right track, I can clean up and squash the commits before merging.

screen shot 2016-03-27 at 17 05 04

File: client/my-sites/upgrades/domain-management/list/item.jsx [hunk 2]
```diff
@@ -24,7 +25,7 @@ const ListItem = React.createClass( {
 
 					<div className="domain-management-list-item__meta">
 						<span className="domain-management-list-item__type">{ this.getDomainTypeText() }</span>
-
+						{ this.props.domain.type !== 'WPCOM' ? this.isDomainExpired( this.props.domain ) : null }
```
Comment (umurkontaci) [inline client/my-sites/upgrades/domain-management/list/item.jsx]: We have a function `isRegisteredDomain` at client/lib/domains/index.js:140

Comment (umurkontaci) [inline client/my-sites/upgrades/domain-management/list/item.jsx]: `isDomainExpired` gives a sense that it will return a boolean, however it returns a component, maybe we should rename it.

Also, no need to use `if ? fn : null`, we can just do `{ isRegisteredDomain( this.props.domain ) && this.domainExpiration() }`

**Comment by umurkontaci** [comment]
Much simpler :).

I made a small comment about code styling and there's also @drewblaisdell 's comment on what happens if it's closer than 24 hour. Once those are fixed this is good to go

File: client/my-sites/upgrades/domain-management/list/item.jsx (hunk 2 above)
Comment (jeremeylduvall) [inline client/my-sites/upgrades/domain-management/list/item.jsx]: @umurkontaci Can definitely change the name. Makes total sense. With `isRegisteredDomain`, that only returns true for registered domains (not mapped, subdomains or site redirects) if I'm reading [`lib/domains/index.js`](https://github.com/Automattic/wp-calypso/blob/f6aa008b7dd1a8c2357626966d1ff6f67c806c0c/client/lib/domains/index.js#L140) correctly. I can use the companion [`isMapped`](https://github.com/Automattic/wp-calypso/blob/f6aa008b7dd1a8c2357626966d1ff6f67c806c0c/client/lib/domains/index.js#L122) and [`isSubdomain`](https://github.com/Automattic/wp-calypso/blob/f6aa008b7dd1a8c2357626966d1ff6f67c806c0c/client/lib/domains/index.js#L109), but that still leaves site redirects. Feels simpler to just filter out `domain.type === WPCOM`. Thoughts?

**Comment by jeremeylduvall** [comment]
I have this reworked with `showDomainExpirationWarning` in place of `isDomainExpired` so it's more clear what the function is actually doing. I'm still using the `domain.type === WPCOM` comparison instead of `isRegisteredDomain` per [the comment here](https://github.com/Automattic/wp-calypso/pull/2728#discussion_r57626309).

> I made a small comment about code styling and there's also drewblaisdell 's comment on what happens if it's closer than 24 hour.

Perhaps I'm misunderstanding, but I don't think this is a problem now that domain.expirationMoment includes a time component like `2016-04-26T00:00:00+00:00`. [Originally](https://github.com/Automattic/wp-calypso/pull/2728#issue-128457971), there was just a date component so the `.startOf( 'day' )` dance was necessary to get the correct value. Now, `domain.expirationMoment.fromNow()` should return the correct value even within 24 hours of expiration. For example, currently, it reads "Expires in 39 minutes" on an example domain:

screen shot 2016-03-28 at 5 21 08 pm

Note, this matches how we perform the check in [`purchases/list/item/`](https://github.com/Automattic/wp-calypso/blob/master/client/me/purchases/list/item/index.jsx#L49), which is what I'm trying to match so it's possible that is clouding my thinking :) .

**Comment by klimeryk** [comment]
I've rebased with `master`, and fixed two bugs I've found:
* [`this.moment()` is not a function](https://github.com/Automattic/wp-calypso/pull/2728/commits/375cd18b19066cd40878a15ce45a762d6276f92e#diff-e22c55909341e15e4352a580bbfe5454R119), you probably wanted to import `moment`.
* [we need to check if `expirationMoment` is not null](https://github.com/Automattic/wp-calypso/pull/2728/commits/375cd18b19066cd40878a15ce45a762d6276f92e#diff-e22c55909341e15e4352a580bbfe5454R119), because for VIP sites it will be `null`.

This LGTM. I'm marking it as such - feel free to test it yourself and merge it :+1:

**Comment by jeremeylduvall** [comment]
`this.moment() is not a function, you probably wanted to import moment.`

Blast—I thought it was failing because `moment` wasn't imported into the test at [`list/test/index.js`](https://github.com/Automattic/wp-calypso/blob/fix/2422-domain-notice/client/my-sites/upgrades/domain-management/list/test/index.js) so I was stumped why that didn't fix the issue. Thanks for fixing! Will test this out and merge.

**Comment by umurkontaci** [comment]
> this.moment() is not a function, you probably wanted to import moment.

`this.moment()` is actually the correct usage since it has the localization information. We inject a mixin that includes `moment` function in every component in React.

Unit tests will throw because the function is not injected and needs to injected manually:

First option: 
`componentClass.__reactAutoBindMap.moment = fn`
This can be done after importing the component.

Second option:
`require('react/lib/ReactInjection').Class.injectMixin({moment: fn})`)
This must be done before important the component and applies to all react components.

[Similar example](test/test/helpers/use-i18n/README.md)

The only usage of imported `moment` module is creating a date 30 days ahead so it does not change the outcome, but since the suggested usage of `moment` is not clear to all of us, this might set a bad precedent.

**Comment by klimeryk** [comment]
Thanks, @umurkontaci, that's very useful information. I didn't know we injected `moment`. I've tried your second approach (since in the test we were already injecting `i18n`).
I've changed the status back to `Needs Review` - please check if everything looks OK to you :)

**Comment by umurkontaci** [comment]
LGTM @klimeryk go for it

@jeremeylduvall thanks a lot for your effort and patience :+1:
